subsys/net/lib/lwm2m/lwm2m_rw_link_format.c

```diff
@@ -142,51 +142,114 @@ static int put_corelink_dimension(struct lwm2m_output_context *out,
 	return len;
 }
 
-static int put_corelink_attributes(struct lwm2m_output_context *out,
-				   const void *ref, uint8_t *buf,
-				   uint16_t buflen)
+static int put_attribute(struct lwm2m_output_context *out,
+			 struct lwm2m_attr *attr, uint8_t *buf,
+			 uint16_t buflen)
 {
-	struct lwm2m_attr *attr = NULL;
 	int used, ret;
-	int len = 0;
+	const char *name = lwm2m_engine_get_attr_name(attr);
 
-	while ((attr = lwm2m_engine_get_next_attr(ref, attr)) != NULL) {
-		const char *name = lwm2m_engine_get_attr_name(attr);
+	if (name == NULL) {
+		/* Invalid attribute, ignore. */
+		return 0;
+	}
 
-		if (name == NULL) {
-			/* Invalid attribute, ignore. */
-			continue;
+	if (attr->type <= LWM2M_ATTR_PMAX) {
+		used = snprintk(buf, buflen, ";%s=%d", name, attr->int_val);
+	} else {
+		uint8_t float_buf[32];
+
+		used = lwm2m_ftoa(&attr->float_val, float_buf,
+				  sizeof(float_buf), 4);
+		if (used < 0 || used >= sizeof(float_buf)) {
+			return -ENOMEM;
 		}
 
-		if (attr->type <= LWM2M_ATTR_PMAX) {
-			used = snprintk(buf, buflen, ";%s=%d", name, attr->int_val);
-		} else {
-			uint8_t float_buf[32];
+		used = snprintk(buf, buflen, ";%s=%s", name, float_buf);
+	}
 
-			used = lwm2m_ftoa(&attr->float_val, float_buf,
-					  sizeof(float_buf), 4);
-			if (used < 0 || used >= sizeof(float_buf)) {
-				return -ENOMEM;
-			}
+	if (used < 0 || used >= buflen) {
+		return -ENOMEM;
+	}
 
-			used = snprintk(buf, buflen, ";%s=%s", name, float_buf);
-		}
+	ret = buf_append(CPKT_BUF_WRITE(out->out_cpkt), buf, used);
+	if (ret < 0) {
+		return ret;
+	}
 
-		if (used < 0 || used >= buflen) {
-			return -ENOMEM;
-		}
+	return ret;
+}
 
-		len += used;
+static int put_attributes(struct lwm2m_output_context *out,
+			  struct lwm2m_attr **attrs, uint8_t *buf,
+			  uint16_t buflen)
+{
+	int ret;
+	int len = 0;
+
+	for (int i = 0; i < NR_LWM2M_ATTR; i++) {
+		if (attrs[i] == NULL) {
+			continue;
+		}
 
-		ret = buf_append(CPKT_BUF_WRITE(out->out_cpkt), buf, used);
+		ret = put_attribute(out, attrs[i], buf, buflen);
 		if (ret < 0) {
 			return ret;
 		}
+
+		len += ret;
 	}
 
 	return len;
 }
 
+static void get_attributes(const void *ref, struct lwm2m_attr **attrs)
+{
+	struct lwm2m_attr *attr = NULL;
+
+	while ((attr = lwm2m_engine_get_next_attr(ref, attr)) != NULL) {
+		if (attr->type >= NR_LWM2M_ATTR) {
+			continue;
+		}
+
+		attrs[attr->type] = attr;
+	}
+}
+
+static int put_corelink_attributes(struct lwm2m_output_context *out,
+				   const void *ref, uint8_t *buf,
+				   uint16_t buflen)
+{
+	struct lwm2m_attr *attrs[NR_LWM2M_ATTR] = { 0 };
+
+	get_attributes(ref, attrs);
+
+	return put_attributes(out, attrs, buf, buflen);
+}
+
+/* Resource-level attribute request - should propagate attributes from Object
+ * and Object Instance.
+ */
+static int put_corelink_attributes_resource(struct lwm2m_output_context *out,
```

Comment:
1.1 makes it possible to have resource instance-level attributes?

Reply:
Yes, see chapter 6.3.4 of the LwM2M 1.1 specification.